### PR TITLE
Allow you to tab through modal text box

### PR DESCRIPTION
**Pull request by itsmichaelwang**
### Description of the Change
We have a modal with elements (1) text box, (2) text box, (3) cancel button, and (4) clone button, respectively, from the top-left to the bottom-right.
screen shot 2017-11-08 at 12 43 16 am

For this modal element, the tabIndexes are -1, -1, 3, 4, respectively. This causes the tabbing order to be (1) > (3) > (4) or (2) > (3) > (4), and then looping forever between (3) and (4), because tabIndex=-1 signifies to skip tabbing for that element. It would appear the root cause is the tabIndex for those elements is hardcoded to be -1. We should probably change that.

Instead, allow a tabIndex to be inferred from the constructor. Now the modal has tabIndexes of 1, 2, 3, 4, as it was intended to be, from the atom github repo.

But now there is a new issue. We can only tab from (1) > (2), but never to (3) or (4). I have tried to understand why this is happening, but I can't come to a conclusion after playing around with the values, nor do I understand how to debug tabbing better than the trial and error approach I have now. Seems like something that's probably better for another ticket, because the tabIndexes are correct now.

### Alternate Designs
Maybe instead of inferring tabIndex from `this.props.element.tabIndex` in line 173, we should instead have some kind of explicit field to specify it, like `this.props.tabIndex`. I would not be opposed to that. Of course, then we need to do some more refactoring work to make tabIndex work for all objects that create this modal. It's a tradeoff between time investment and tech debt.

Also, I'm not sure if this would break other things. Seems like a pretty broad change.

### Why Should This Be In Core?
The core issue here is that the tabIndex is being overwritten in `text-editor-component.js` from a hardcoded value. Even if another package (in this case, the Github package) specifies a tabIndex, it will be squashed by what's hardcoded, and that seems wrong.

### Benefits
You can tab through your text boxes now.

### Possible Drawbacks
You can't tab through your buttons now. I think you also can't use the escape button to exit out of a modal, but you couldn't do that anyways unless the focus was on the "Cancel" button, so the problems are linked.

I think the tabbing thing is actually a separate issue, and this PR just solves the tabIndex being squashed by a hardcoded value.

### Applicable Issues
#16004

### CODE REVIEW

**Comment by itsmichaelwang** [comment]
I will address the failing tests first.

**Comment by itsmichaelwang** [comment]
@50Wliu unit tests passed

**Comment by akonwi** [comment]
bump

File: src/text-editor-component.js [hunk 1]
```diff
@@ -3586,7 +3587,7 @@ class CursorsAndInputComponent {
         compositionupdate: didCompositionUpdate,
         compositionend: didCompositionEnd
       },
-      tabIndex: -1,
+      tabIndex: this.tabIndex,
```
Comment (jasonrudolph) [inline src/text-editor-component.js]: I think this is an attempt to access the [`tabIndex` property on the `TextEditorComponent` object](https://github.com/atom/atom/blob/df2f3c3124a6e7ea7f9e23c912cff9abfefda0b9/src/text-editor-component.js#L173). However, we're inside the [`CursorsAndInputComponent` class](https://github.com/atom/atom/blob/df2f3c3124a6e7ea7f9e23c912cff9abfefda0b9/src/text-editor-component.js#L3494-L3604) here, and the `CursorsAndInputComponent` objects don't currently have a `tabIndex` property. As a result, `this.tabIndex` is always `undefined`, and that results in the `input` element getting its tabIndex set to `0`. I _think_ this is the cause of the issues you noted in the [possible drawbacks section](https://github.com/atom/atom/pull/16127#issue-151324878).

I've attempted to resolve those issues in #16492. If you have a chance to look at that pull request, I'd love to have your feedback. :bow:

**Comment by jasonrudolph** [comment]
I cherry-picked the commits from this pull request into #16492, and I added some additional commits to 1) address the drawbacks noted in the pull request body above and 2) add tests for this new behavior.

Since #16492 is merged now, and since it includes the commits from this pull request, I'm going to close this pull request. These improvements improvements will likely be available in Atom 1.25.

@itsmichaelwang: Thanks again for this contribution!

**Pull request closed without merging**